src/python/pants/backend/python/mixed_interpreter_constraints/py_constraints.py

```diff
@@ -1,35 +1,152 @@
+    console: Console,
+    py_constraints_subsystem: PyConstraintsSubsystem,
+    python_setup: PythonSetup,
 ) -> PyConstraintsGoal:
+    if py_constraints_subsystem.summary:
+        if addresses:
+            console.print_stderr(
+                "The `py-constraints --summary` goal does not take file/target arguments. It will "
+                "generate a summary of your entire repository. You can then use Google "
+                "Sheets/Excel to post-process/filter the data."
+            )
+            return PyConstraintsGoal(exit_code=1)
+
+        all_expanded_targets, all_explicit_targets = await MultiGet(
+            Get(Targets, AddressSpecs([DescendantAddresses("")])),
+            Get(UnexpandedTargets, AddressSpecs([DescendantAddresses("")])),
+        )
+        all_python_targets = sorted(
+            {
+                t
+                for t in (*all_expanded_targets, *all_explicit_targets)
+                if t.has_field(InterpreterConstraintsField)
+                or t.has_field(PythonInterpreterCompatibility)
+            },
+            key=lambda tgt: tgt.address,
+        )
+
+        constraints_per_tgt = [
+            PexInterpreterConstraints.create_from_targets([tgt], python_setup)
+            for tgt in all_python_targets
+        ]
+
+        transitive_targets_per_tgt = await MultiGet(
+            Get(TransitiveTargets, TransitiveTargetsRequest([tgt.address]))
+            for tgt in all_python_targets
+        )
+        transitive_constraints_per_tgt = [
+            PexInterpreterConstraints.create_from_targets(transitive_targets.closure, python_setup)
+            for transitive_targets in transitive_targets_per_tgt
+        ]
+
+        dependees_per_root = await MultiGet(
+            Get(Dependees, DependeesRequest([tgt.address], transitive=True, include_roots=False))
+            for tgt in all_python_targets
+        )
+
+        data = [
+            {
+                "Target": tgt.address.spec,
+                "Constraints": str(constraints),
+                "Transitive Constraints": str(transitive_constraints),
+                "# Dependencies": len(transitive_targets.dependencies),
+                "# Dependees": len(dependees),
+            }
+            for tgt, constraints, transitive_constraints, transitive_targets, dependees in zip(
+                all_python_targets,
+                constraints_per_tgt,
+                transitive_constraints_per_tgt,
+                transitive_targets_per_tgt,
+                dependees_per_root,
+            )
+        ]
+
+        output_stream = StringIO()
+        writer = csv.DictWriter(
+            output_stream,
+            fieldnames=[
+                "Target",
+                "Constraints",
+                "Transitive Constraints",
+                "# Dependencies",
+                "# Dependees",
+            ],
+        )
+        writer.writeheader()
+        for entry in data:
+            writer.writerow(entry)
```

Comment:
AFAIK, writing to the `Console` should disable all other output for the length of the run, so it should not be necessary to write this to a buffer and then flush at the end like this. Should be able to just write line at a time to the `Console`.